pkg/services/datasources/service/secrets_mig.go

```diff
@@ -0,0 +1,52 @@
+package service
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/services/datasources"
+)
+
+type DataSourceSecretMigrationService struct {
+	dataSourcesService datasources.DataSourceService
+}
+
+func ProvideDataSourceMigrationService(dataSourcesService datasources.DataSourceService) *DataSourceSecretMigrationService {
+	return &DataSourceSecretMigrationService{
+		dataSourcesService: dataSourcesService,
+	}
+}
+
+func (s *DataSourceSecretMigrationService) Migrate(ctx context.Context) error {
+	query := &datasources.GetAllDataSourcesQuery{}
+	err := s.dataSourcesService.GetAllDataSources(ctx, query)
```

Comment:
As discussed, this will run every time a Grafana instance starts up and you'll have to iterate all datasources to check `secretMigrationComplete`. In a small installation this is probably not a problem, but might be non-optimal in larger installations (thousands of datasources). With my suggestion of using kvstore to store migration state might be able to skip iterating all datasources.

Reply:
Will resolve this conversation in favor of keeping discussion on https://github.com/grafana/grafana/pull/50463#discussion_r913493299